### PR TITLE
LPS-164940 | Expose ERC for KB Attachment

### PR DESCRIPTION
**Pull request by Hijky**
**Background:**
Add a testcase to expose ERC for KB Attachment

**Test Plan**
**Background**
Given with postSiteKnowledgeBaseArticle and siteId to create a knowledge base article with externalReferenceCode
And Given with postKnowledgeBaseArticleKnowledgeBaseAttachment and knowledgeBaseArticleId to upload an attachment with externalReferenceCode to knowledge base article
And Given with postKnowledgeBaseArticleKnowledgeBaseArticle and parentKnowledgeBaseArticleId to create a child knowledge base article
And Given with postSiteKnowledgeBaseFolder and siteId to create a knowledgeBaseFolder
And Given with postKnowledgeBaseFolderKnowledgeBaseFolder and parentKnowledgeBaseFolderId to create a folder inside parent folder

CanCreateAttachmentWithDefaultExternalReferenceCodeForKnowledgeBaseArticle
When with postKnowledgeBaseArticleKnowledgeBaseAttachment and knowledgeBaseArticleId to upload an attachment to knowledge base article
Then the knowledgeBaseAttachment with externalReferenceCode is created

CanCreateAttachmentWithExternalReferenceCodeForChildKnowledgeBaseArticle
When with postKnowledgeBaseArticleKnowledgeBaseAttachment and knowledgeBaseArticleId to upload an attachment with custom externalReferenceCode to the child knowledge base article
Then the knowledgeBaseAttachment with custom externalReferenceCode is created

CanCreateAttachmentWithExternalReferenceCodeForKnowledgeBaseArticleInFolder
And Given with postKnowledgeBaseFolderKnowledgeBaseArticle and knowledgeBaseFolderId to create a knowledge base article with externalReferenceCode
When with postKnowledgeBaseArticleKnowledgeBaseAttachment and knowledgeBaseArticleId to upload an attachment with externalReferenceCode to knowledge base article
Then the knowledgeBaseAttachment with custom externalReferenceCode is created

CanCreateAttachmentWithExternalReferenceCodeForKnowledgeBaseArticleInChildFolder
And Given with postKnowledgeBaseFolderKnowledgeBaseArticle and knowledgeBaseFolderId to create a knowledge base article with externalReferenceCode inside child folder
When with postKnowledgeBaseArticleKnowledgeBaseAttachment and knowledgeBaseArticleId to upload an attachment with externalReferenceCode to knowledge base article
Then the knowledgeBaseAttachment with custom externalReferenceCode is created

CanGetAttachmentForKnowledgeBaseArticleByExternalReferenceCode
When with getSiteKnowledgeBaseArticleByExternalReferenceCodeKnowledgeBaseArticleExternalReferenceCodeKnowledgeBaseAttachmentByExternalReferenceCode and knowledgeBaseExternalReferenceCode and knowledgeBaseAttachmentExternalReferenceCode to retrieve the knwoledgeBaseAttachment
Then I can see details of the knowledgeBaseAttachment

CanGetAttachmentForChildKnowledgeBaseArticleByExternalReferenceCode
And Given with postKnowledgeBaseArticleKnowledgeBaseAttachment and knowledgeBaseArticleId to upload an attachment with externalReferenceCode to the child knowledge base article
When with getSiteKnowledgeBaseArticleByExternalReferenceCodeKnowledgeBaseArticleExternalReferenceCodeKnowledgeBaseAttachmentByExternalReferenceCode and knowledgeBaseExternalReferenceCode and knowledgeBaseAttachmentExternalReferenceCode to retrieve the attachment for child knowledgeBaseArticle
Then I can see details of the knowledgeBaseAttachment

CanGetAttachmentForKnowledgeBaseArticleInFolderByExternalReferenceCode
And Given with postKnowledgeBaseFolderKnowledgeBaseArticle and knowledgeBaseFolderId to create a knowledge base article with externalReferenceCode
And Given with postKnowledgeBaseArticleKnowledgeBaseAttachment and knowledgeBaseArticleId to upload an attachment with externalReferenceCode to knowledge base article in folder
When with getSiteKnowledgeBaseArticleByExternalReferenceCodeKnowledgeBaseArticleExternalReferenceCodeKnowledgeBaseAttachmentByExternalReferenceCode and knowledgeBaseExternalReferenceCode and knowledgeBaseAttachmentExternalReferenceCode to retrieve the attachment for knowledgeBaseArticle in folder
Then I can see details of the knowledgeBaseAttachment

CanGetAttachmentForKnowledgeBaseArticleInChildFolderByExternalReferenceCode
And Given with postKnowledgeBaseFolderKnowledgeBaseArticle and knowledgeBaseFolderId to create a knowledge base article with externalReferenceCode inside child folder
And Given with postKnowledgeBaseArticleKnowledgeBaseAttachment and knowledgeBaseArticleId to upload an attachment with externalReferenceCode to knowledge base article inside child folder
When with getSiteKnowledgeBaseArticleByExternalReferenceCodeKnowledgeBaseArticleExternalReferenceCodeKnowledgeBaseAttachmentByExternalReferenceCode and knowledgeBaseExternalReferenceCode and knowledgeBaseAttachmentExternalReferenceCode to retrieve the attachment for knowledgeBaseArticle inside child folder
Then I can see details of the knowledgeBaseAttachment

CanDeleteAttachementForKnowledgeBaseArticleByExternalReferenceCode
When with deleteSiteKnowledgeBaseArticleByExternalReferenceCodeKnowledgeBaseArticleExternalReferenceCodeKnowledgeBaseAttachmentByExternalReferenceCode and siteId and knowledgeBaseExternalReferenceCode and knowledgeBaseAttachmentExternalReferenceCode to delete the knwoledgeBaseAttachment Then the knowledgeBaseAttachment is deleted

CanDeleteAttachmentForKnowledgeBaseArticleInFolderByExternalReferenceCode
And Given with postKnowledgeBaseFolderKnowledgeBaseArticle and knowledgeBaseFolderId to create a knowledge base article with externalReferenceCode And Given with postKnowledgeBaseArticleKnowledgeBaseAttachment and knowledgeBaseArticleId to upload an attachment with externalReferenceCode to knowledge base article in folder When with deleteSiteKnowledgeBaseArticleByExternalReferenceCodeKnowledgeBaseArticleExternalReferenceCodeKnowledgeBaseAttachmentByExternalReferenceCode and siteId and knowledgeBaseExternalReferenceCode and knowledgeBaseAttachmentExternalReferenceCode to delete the attachment for knowledgeBaseArticle in folder Then the knowledgeBaseAttachment in folder is deleted

**Jira ticket:**
https://issues.liferay.com/browse/LPS-164940

### CODE REVIEW

**Comment by Hijky** [comment]
ci:test:headless-functional:nocompile

**Review by magjed4289 — APPROVED** [review]

**Comment by magjed4289** [comment]
@Hijky since it is  `testray.main.component.name = "Lima Headless";` we need to run their test suite in order to see whether the changes are passing

**Comment by Hijky** [comment]
ci:test:lima-headless

File: modules/apps/headless/headless-delivery/headless-delivery-test/src/testFunctional/macros/KnowledgeBaseAttachmentAPI.macro [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+				var externalReferenceCode = "";
+			}
+
+			var body = '''
+				-H Content-Type: multipart/form-data \
+				-F file=@${attachmentPath} \
+				-F knowledgeBaseAttachment={"externalReferenceCode": "${externalReferenceCode}"}
+			''';
+
+			var command = StringUtil.add(${command}, ${body}, "");
```
Comment (yanan) [inline modules/apps/headless/headless-delivery/headless-delivery-test/src/testFunctional/macros/KnowledgeBaseAttachmentAPI.macro:35]: Can we return curl here?

File: modules/apps/headless/headless-delivery/headless-delivery-test/src/testFunctional/macros/KnowledgeBaseArticleAPI.macro [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+	macro deleteKnowledgeBaseArticleById {
+		Variables.assertDefined(parameterList = ${knowledgeBaseArticleId});
+
+		var curl = KnowledgeBaseArticleAPI._curlKnowledgeBaseArticle(knowledgeBaseArticleId = ${knowledgeBaseArticleId});
+
+		JSONCurlUtil.delete(${curl});
+	}
+
+	macro setUpGlobalKnowledgeBaseArticleId {
```
Comment (yanan) [inline modules/apps/headless/headless-delivery/headless-delivery-test/src/testFunctional/macros/KnowledgeBaseArticleAPI.macro:64]: setUpGlobalKnowledgeBaseArticleIdAndErc?

**Comment by Hijky** [comment]
ci:test:sf

**Comment by Hijky** [comment]
ci:test:headless-functional:nocompile

**Comment by Hijky** [comment]
ci:test:lima-headless

**Review by yanan — APPROVED** [review]

**Review by magjed4289 — APPROVED** [review]

**Comment by Hijky** [comment]
ci:forward

**Pull request closed without merging**